sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py

```diff
@@ -71,148 +72,154 @@ def __init__(self, eventhub_client: EventHubClient, consumer_group_name: str,
 
     def __repr__(self):
         return 'EventProcessor: id {}'.format(self._id)
 
     async def start(self):
         """Start the EventProcessor.
 
-        1. retrieve the partition ids from eventhubs.
-        2. claim partition ownership of these partitions.
-        3. repeatedly call EvenHubConsumer.receive() to retrieve events and call user defined PartitionProcessor.process_events().
+        1. Calls the OwnershipManager to keep claiming and balancing ownership of partitions in an
+        infinitely loop until self.stop() is called.
+        2. Cancels tasks for partitions that are no longer owned by this EventProcessor
+        3. Creates tasks for partitions that are newly claimed by this EventProcessor
+        4. Keeps tasks running for partitions that haven't changed ownership
+        5. Each task repeatedly calls EvenHubConsumer.receive() to retrieve events and call user defined partition processor
 
         :return: None
 
         """
         log.info("EventProcessor %r is being started", self._id)
-        partition_ids = await self._eventhub_client.get_partition_ids()
-        claimed_list = await self._claim_partitions(partition_ids)
-        await self._start_claimed_partitions(claimed_list)
+        ownership_manager = OwnershipManager(self, self._eventhub_client, self._ownership_timeout)
+        if not self._running:
+            self._running = True
+            while self._running:
+                claimed_ownership_list = await ownership_manager.claim_ownership()
+                if claimed_ownership_list:
+                    claimed_partition_ids = [x["partition_id"] for x in claimed_ownership_list]
+                    to_cancel_list = self._tasks.keys() - claimed_partition_ids
+                    self._create_tasks_for_claimed_ownership(claimed_ownership_list)
+                else:
+                    log.warning("EventProcessor %r hasn't claimed an ownership. It keeps claiming.", self._id)
+                    to_cancel_list = self._tasks.keys()
+                if to_cancel_list:
+                    self._cancel_tasks_for_partitions(to_cancel_list)
+                    log.info("EventProcesor %r has cancelled partitions %r", self._id, to_cancel_list)
+                await asyncio.sleep(self._polling_interval)
 
     async def stop(self):
-        """Stop all the partition consumer
+        """Stop claiming ownership and all the partition consumers owned by this EventProcessor
 
-        This method cancels tasks that are running EventHubConsumer.receive() for the partitions owned by this EventProcessor.
+        This method stops claiming ownership of owned partitions and cancels tasks that are running
+        EventHubConsumer.receive() for the partitions owned by this EventProcessor.
 
         :return: None
 
         """
+        self._running = False
         for i in range(len(self._tasks)):
-            task = self._tasks.pop()
+            task = self._tasks.popitem()[1]
```

Comment:
Purely an alternative style suggestion: 
```
_, task = self._tasks.popitem()
```
For some reason I prefer that to seeing indexing directly on a function call (but that's just me)

Reply:
Since your suggestion looks cooler, I'm going to use it.